fecgov/openFEC
# 4958-upgrade flask

## Summary (required)

- Resolves #4958 

This is a major version upgrade with several breaking changes. Flask-Script has been deprecated, so it was removed and the Flask in-built CLI (which runs with click) was used to replace that functionality. There were a few changes to commands and some extra variables that need to be set which are noted in the README. **They will all be in the how-to below, but note that to run the server you must now use `flask run` or `python cli.py run`**

The commands used to be run with, for example: `python manage.py refresh_materialized`, in my first commit, but if I keep the CLI decorators on those functions I get a TypeError when I run the integration tests with pytest. 
If you run `pytest tests/integration/test_advisory_opinions.py -x` on the original commit you will see the errors. I fixed this issue by re-defining the functions for the CLI in a separate file (cli.py). If anyone has a better solution, or would like to offer suggestions on this item I would be very open to switching it up. 

 I've uploaded a .flaskenv file so we wouldn't have to manually export the FLASK_APP and FLASK_DEBUG variables in terminal. I added python-dotenv to the requirements.txt file which automatically loads the variables from the .flaskenv file. I ran locust tests (results below) and found no difference from our current speeds and failures. 

NOTE: I will update [the wiki](https://github.com/fecgov/openFEC/wiki/Elasticsearch-7.x.0-setup-and-test-locally) before the release is live.

I've noted that we have a ticket to fix the import app anti-pattern and think we should fix that separately. 

Sources I found helpful:
https://flask.palletsprojects.com/en/2.1.x/cli/
https://blog.miguelgrinberg.com/post/migrating-from-flask-script-to-the-new-flask-cli
https://click.palletsprojects.com/en/8.1.x/commands/


# Required reviewers
2-3 devs

## Impacted areas of the application

General components of the application that this PR will affect:

-  Local testing environment (specifically running any task in your terminal)


## How to test

- `gh pr checkout 5186`
- `pyenv install 3.8.12` 
- `pyenv virtualenv 3.8.12 venv-api3812` 
- `pyenv activate venv-api3812`
- `pip install -r requirements.txt` 
- `pip install -r requirements-dev.txt`
- `rm -rf node_modules` (optional step) (check for correct node version, switch with `nvm install 14.15.5`)
- `npm i && npm run build` 
- `dropdb cfdm_test`
- `createdb cfdm_test`
- `invoke create_sample_db`
- `pytest` (all tests should pass)
-  `python cli.py run` OR `flask run` and test swagger-ui
- try out the other CLI commands such as `python cli.py refresh_materialized`,  `python cli.py create_index` (any of the tasks in the wiki, but make sure to replace manage.py with cli.py and also remove flask options ie -i, -f,...would be happy to pair with anyone on this and running tasks)
-  deploy this branch to dev 
-  login to cloud.gov & make sure you're on dev 
- `cf tasks api` to see the tasks that have been tested 
- `cf run-task api --command "python cli.py display_snapshot_detail repository_archived_murs archived_murs*" -m 2G --name display_snapshot_detail" -m 2G --name display_snapshot_detail` (or any other task you'd like to run) 
- `cf tasks api` (Should run successfully---takes a few minutes depending on the command) 
- run locust testing (I already did this and the results are in a comment below.)
